model/historical_test_data.go

```diff
@@ -258,3 +269,484 @@ func (i *HistoricalTestDataInfo) validate() error {
+	Variant  string    `bson:"variant"`
+	Date     time.Time `bson:"date"`
+
+	NumPass         int           `bson:"num_pass"`
+	NumFail         int           `bson:"num_fail"`
+	AverageDuration time.Duration `bson:"avg_duration"`
+	LastUpdate      time.Time     `bson:"last_update"`
+}
+
+var (
+	aggregatedHistoricalTestDataTestNameKey    = bsonutil.MustHaveTag(AggregatedHistoricalTestData{}, "TestName")
+	aggregatedHistoricalTestDataTaskNameKey    = bsonutil.MustHaveTag(AggregatedHistoricalTestData{}, "TaskName")
+	aggregatedHistoricalTestDataVariantKey     = bsonutil.MustHaveTag(AggregatedHistoricalTestData{}, "Variant")
+	aggregatedHistoricalTestDataDateKey        = bsonutil.MustHaveTag(AggregatedHistoricalTestData{}, "Date")
+	aggregatedHistoricalTestDataNumPassKey     = bsonutil.MustHaveTag(AggregatedHistoricalTestData{}, "NumPass")
+	aggregatedHistoricalTestDataNumFailKey     = bsonutil.MustHaveTag(AggregatedHistoricalTestData{}, "NumFail")
+	aggregatedHistoricalTestDataAvgDurationKey = bsonutil.MustHaveTag(AggregatedHistoricalTestData{}, "AverageDuration")
+	aggregatedHistoricalTestDataLastUpdateKey  = bsonutil.MustHaveTag(AggregatedHistoricalTestData{}, "LastUpdate")
+)
+
+// GetHistoricalTestData queries the historical test data using a filter.
+func GetHistoricalTestData(ctx context.Context, env cedar.Environment, filter HistoricalTestDataFilter) ([]AggregatedHistoricalTestData, error) {
+	err := filter.validate()
+	if err != nil {
+		return nil, errors.Wrap(err, "the provided HistoricalTestDataFilter is invalid")
+	}
+
+	var data []AggregatedHistoricalTestData
+	pipeline := filter.queryPipeline()
+	cursor, err := env.GetDB().Collection(historicalTestDataCollection).Aggregate(ctx, pipeline)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem aggregating test data")
+	}
+	if err = cursor.All(ctx, &data); err != nil {
+		return nil, errors.Wrap(err, "problem unmarshaling aggregated test data")
+	}
+
+	return data, nil
+}
+
+const htdMaxQueryLimit = 1001
+
+// HTDGroupBy represents the possible groupings of historical test data.
+type HTDGroupBy string
+
+const (
+	HTDGroupByTest    HTDGroupBy = "test"
+	HTDGroupByTask    HTDGroupBy = "task"
+	HTDGroupByVariant HTDGroupBy = "variant"
+)
+
+func (gb HTDGroupBy) validate() error {
+	switch gb {
+	case HTDGroupByVariant:
+	case HTDGroupByTask:
+	case HTDGroupByTest:
+	default:
+		return errors.Errorf("invalid HTDGroupBy value: %s", gb)
+	}
+
+	return nil
+}
+
+// HTDSort represents the sorting options for historical test data.
+type HTDSort string
+
+const (
+	HTDSortEarliestFirst HTDSort = "earliest"
+	HTDSortLatestFirst   HTDSort = "latest"
+)
+
+func (s HTDSort) validate() error {
+	switch s {
+	case HTDSortEarliestFirst:
+	case HTDSortLatestFirst:
+	default:
+		return errors.Errorf("invalid HTDSort value: %s", s)
+	}
+
+	return nil
+}
+
+// HTDStartAt represents parameters that allow a search query to resume at a
+// specific point. Used for pagination.
```

Comment:
This seems to have an encoded assumption that it's hierarchical, as in, if you specify a variant, you must specify task and test. Can we just add a short comment on this so anyone who reads this has less confusion next time?

Reply:
So test must always be specified (I think that is a given since this is the historical test data), but the only time variant and task must be specified is if `HTDGroupByVariant` is set (then variant and task must be set) or `HTDGroupByTask` is set (in which case task must be set).